cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -350,20 +353,37 @@ public ApplicationWithPrograms deployApp(Id.Namespace namespace, @Nullable Strin
   public void removeAll(final Id.Namespace namespaceId) throws Exception {
     List<ApplicationSpecification> allSpecs = new ArrayList<>(
       store.getAllApplications(namespaceId));
-
     //Check if any program associated with this namespace is running
-    boolean appRunning = runtimeService.checkAnyRunning(new Predicate<Id.Program>() {
+    Iterable<ProgramRuntimeService.RuntimeInfo> runtimeInfos =
+      Iterables.filter(runtimeService.listAll(ProgramType.values()),
+                       new Predicate<ProgramRuntimeService.RuntimeInfo>() {
       @Override
-      public boolean apply(Id.Program programId) {
-        return programId.getApplication().getNamespace().equals(namespaceId);
+      public boolean apply(ProgramRuntimeService.RuntimeInfo runtimeInfo) {
+        return runtimeInfo.getProgramId().toEntityId().getNamespace().equals(namespaceId.getId());
       }
-    }, ProgramType.values());
+    });
 
-    if (appRunning) {
-      throw new CannotBeDeletedException(namespaceId, "One of the program associated with this namespace is still " +
-        "running");
+    Set<ProgramId> runningPrograms = new HashSet<>();
+    for (ProgramRuntimeService.RuntimeInfo runtimeInfo : runtimeInfos) {
+      runningPrograms.add(runtimeInfo.getProgramId().toEntityId());
     }
 
+    if (!runningPrograms.isEmpty()) {
+      String appAllRunningPrograms = Joiner.on(',')
+        .join(Iterables.transform(runningPrograms, new Function<ProgramId, String>() {
+          @Override
+          public String apply(ProgramId program) {
+            return program.getApplication() + ": " + program.getProgram();
+          }
+        }));
+      throw new CannotBeDeletedException(namespaceId,
+                                         "The following programs are still running: " + appAllRunningPrograms) {
+        @Override
+        public int getStatusCode() {
+          return HttpResponseStatus.FORBIDDEN.getCode();
```

Comment:
Can you add the comment about why it is FORBIDDEN instead of CONFLICT? It was "// Keeping this for backward compatibility. Ideally this should return conflict, not forbidden.".